### ISSUE TITLE
KMS Key With Vulnerable Policy

### ISSUE THREAD

**Issue by ccsalway** [Checkmarx/kics#4779]
I get the following HIGH level error when using the KeyPolicy as prescribed by AWS when creating a KMS key....

```
KMS Key With Vulnerable Policy, Severity: HIGH, Results: 1
Description: Checks if the policy is vulnerable and needs updating.
Platform: Terraform

        [1]: kms.tf:5

                005:   policy = jsonencode({
                006:     "Version" : "2012-10-17",
```

```
"expected_value": "aws_kms_key[master].policy does not have wildcard in 'Action' or 'Principal'",
"actual_value": "aws_kms_key[master].policy has wildcard in 'Action' or 'Principal'"
```

```
resource "aws_kms_key" "master" {
  description             = "Managed by Terraform"
  enable_key_rotation     = true
  policy = jsonencode({
    "Version" : "2012-10-17",
    "Statement" : [
      {
        "Sid" : "Enable IAM User Permissions",
        "Effect" : "Allow",
        "Principal" : {
          "AWS" : "arn:aws:iam::${local.aws.account_id}:root"
        },
        "Action" : "kms:*",
        "Resource" : "*"
      }
   ]
  })
}
```

Coincedentially, If I leave off the policy from the terraform above - AWS will apply the same (default) policy - KICS doesnt complain.

In retrospect, the IAM policy to prevent this would be

```
{
    "Version": "2012-10-17",
    "Statement": [
        {
            "Effect": "Allow",
            "Action": [
                "kms:EnableKey",
                "kms:GetPublicKey",
                "kms:ImportKeyMaterial",
                "kms:Decrypt",
                "kms:UntagResource",
                "kms:GenerateRandom",
                "kms:PutKeyPolicy",
                "kms:GenerateDataKeyWithoutPlaintext",
                "kms:Verify",
                "kms:ListResourceTags",
                "kms:CancelKeyDeletion",
                "kms:GenerateDataKeyPair",
                "kms:GetParametersForImport",
                "kms:DescribeCustomKeyStores",
                "kms:SynchronizeMultiRegionKey",
                "kms:DeleteCustomKeyStore",
                "kms:TagResource",
                "kms:UpdatePrimaryRegion",
                "kms:UpdateCustomKeyStore",
                "kms:Encrypt",
                "kms:GetKeyRotationStatus",
                "kms:ScheduleKeyDeletion",
                "kms:ReEncryptTo",
                "kms:DescribeKey",
                "kms:ConnectCustomKeyStore",
                "kms:Sign",
                "kms:CreateGrant",
                "kms:EnableKeyRotation",
                "kms:ListKeyPolicies",
                "kms:UpdateKeyDescription",
                "kms:ListRetirableGrants",
                "kms:GetKeyPolicy",
                "kms:DeleteImportedKeyMaterial",
                "kms:GenerateDataKeyPairWithoutPlaintext",
                "kms:DisableKey",
                "kms:ReEncryptFrom",
                "kms:DisableKeyRotation",
                "kms:RetireGrant",
                "kms:ListGrants",
                "kms:UpdateAlias",
                "kms:ListKeys",
                "kms:RevokeGrant",
                "kms:ListAliases",
                "kms:GenerateDataKey",
                "kms:CreateAlias",
                "kms:DisconnectCustomKeyStore",
                "kms:DeleteAlias"
            ],
            "Resource": "*"
        }
    ]
}
```

**Comment by rafaela-soares** [comment]
Hi, @ccsalway 🙂 Thank you for getting in touch!

Such great observations! We will take a look!

**Comment by rafaela-soares** [comment]
Hello, @ccsalway!

The query `KMS Key With Vulnerable Policy` considers vulnerable every policy with a wildcard in `Principal` and/or `Action`.  So, it recommends not defining it.

In fact, you are right. The AWS documentation presents your third snippet code as the default key policy for KMS keys. However, the default policy grants all the KMS actions and KICS recommends granting the least privileges possible.

We added an extra rule to verify if the `policy` is undefined in PR #4811. And now we also verify if there is no condition defined in the policy statement.

Regarding your last sniped code, since it is an edge case, we decided not to cover it for now.

If you have further questions, let me know 🙂  Thank you so much for your observations! 🚀
